pvlib/iam.py

```diff
@@ -940,3 +941,360 @@ def schlick_diffuse(surface_tilt):
         cug = pd.Series(cug, surface_tilt.index)
 
     return cuk, cug
+
+
+def _get_model(model_name):
+    # check that model is implemented
+    model_dict = {'ashrae': ashrae, 'martin_ruiz': martin_ruiz,
+                  'physical': physical}
+    try:
+        model = model_dict[model_name]
+    except KeyError:
+        raise NotImplementedError(f"The {model_name} model has not been \
+                                    implemented")
+
+    return model
+
+
+def _check_params(model_name, params):
+    # check that the parameters passed in with the model
+    # belong to the model
+    exp_params = _IAM_MODEL_PARAMS[model_name]
+    if set(params.keys()) != exp_params:
+        raise ValueError(f"The {model_name} model was expecting to be passed \
+                         {', '.join(list(exp_params))}, but \
+                         was handed {', '.join(list(params.keys()))}")
+
+
+def _sin_weight(aoi):
+    return 1 - sind(aoi)
+
+
+def _residual(aoi, source_iam, target, target_params,
+              weight_function=_sin_weight,
+              weight_args=None):
+    # computes a sum of weighted differences between the source model
+    # and target model, using the provided weight function
+
+    if weight_args is None:
+        weight_args = {}
+
+    weight = weight_function(aoi, **weight_args)
+
+    # if aoi contains values outside of interval (0, 90), annihilate
+    # the associated weights (we don't want IAM values from AOI outside
+    # of (0, 90) to affect the fit; this is a possible issue when using
+    # `iam.fit`, but not an issue when using `iam.convert`, since in
+    # that case aoi is defined internally)
+    weight = weight * np.logical_and(aoi >= 0, aoi <= 90).astype(int)
+
+    diff = np.abs(source_iam - np.nan_to_num(target(aoi, *target_params)))
```

Comment:
Thinking more about user-defined functions, I think it's worth considering allowing the user to define the entire objective function instead of just the weight component of it.  Of course motivated users can always write their own versions of these functions, but it struck me as strange to allow custom weights but hard-code the rest of the objective function.